### PR TITLE
Restore `GT_INDEX_ADDR` after #13682.

### PR DESCRIPTION
**Pull request by pgavlin**
These changes undo the reversion of the `GT_INDEX_ADDR` changes and fix a few bugs that were identified after the fact. Each bug fix is contained in a separate commit.

Supersedes #13639 and #13665.

Fixes https://github.com/dotnet/corefx/issues/23586.

### CODE REVIEW

**Comment by pgavlin** [comment]
@briansull @BruceForstall @CarolEidt @jkotas @dotnet/jit-contrib PTAL

**Comment by pgavlin** [comment]
@dotnet-bot help

**Comment by pgavlin** [comment]
@dotnet-bot 
test OSX10.12 gcstress0x3
test Windows_NT jitstressregs0x10
test Windows_NT jitstress2

**Comment by pgavlin** [comment]
@dotnet-bot test Windows_NT x64 corefx_minopts

File: src/jit/gentree.cpp [hunk 1]
```diff
@@ -299,6 +299,7 @@ void GenTree::InitNodeSize()
     GenTree::s_gtNodeSizes[GT_FTN_ADDR]         = TREE_NODE_SZ_LARGE;
     GenTree::s_gtNodeSizes[GT_BOX]              = TREE_NODE_SZ_LARGE;
     GenTree::s_gtNodeSizes[GT_INDEX]            = TREE_NODE_SZ_LARGE;
+    GenTree::s_gtNodeSizes[GT_INDEX_ADDR]        = TREE_NODE_SZ_LARGE;
```
Comment (briansull) [inline src/jit/gentree.cpp]: Extra space here

**Comment by pgavlin** [comment]
@dotnet-bot test OSX10.12 gcstress0x3

**Comment by pgavlin** [comment]
CoreFX minopts failure is a build issue.

File: src/jit/compiler.h [hunk 2]
```diff
@@ -9329,6 +9331,26 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
         return compRoot->m_arrayInfoMap;
     }
 
+    inline bool TryGetArrayInfo(GenTreeIndir* indir, ArrayInfo* arrayInfo)
```
Comment (briansull) [inline src/jit/compiler.h]: This method could use a function header explaining what it handles. (why it returns true vs. false etc...)

File: src/jit/gentree.cpp [hunk 3]
```diff
@@ -7569,14 +7611,14 @@ GenTreePtr Compiler::gtCloneExpr(
             case GT_IND:
             case GT_OBJ:
             case GT_STORE_OBJ:
-                if (tree->gtFlags & GTF_IND_ARR_INDEX)
+            {
+                ArrayInfo arrInfo;
+                if (TryGetArrayInfo(tree->AsIndir(), &arrInfo) && !tree->AsIndir()->gtOp1->OperIs(GT_INDEX_ADDR))
```
Comment (briansull) [inline src/jit/gentree.cpp]: Did you consider passing a argument stating whether we should cons up an ArrayInfo for the GT_INDEX_ADDR case.
Here we cons it up and discard it.

File: src/jit/compiler.h [hunk 4]
```diff
@@ -9329,6 +9331,26 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
...
+    {
+        if ((indir->gtFlags & GTF_IND_ARR_INDEX) == 0)
+        {
+            return false;
+        }
+
+        if (indir->gtOp1->OperIs(GT_INDEX_ADDR))
+        {
+            GenTreeIndexAddr* const indexAddr = indir->gtOp1->AsIndexAddr();
+            *arrayInfo = ArrayInfo(indexAddr->gtElemType, indexAddr->gtElemSize, indexAddr->gtElemOffset,
```
Comment (briansull) [inline src/jit/compiler.h]: Can you add a comment saying why we cons up an ArrayInfo here.

File: src/jit/codegenarmarch.cpp [hunk 5]
```diff
@@ -1540,6 +1544,101 @@ void CodeGen::genCodeForLclFld(GenTreeLclFld* tree)
...
+{
+    GenTree* const base  = node->Arr();
+    GenTree* const index = node->Index();
+
+    genConsumeReg(base);
+    genConsumeReg(index);
+
+    // NOTE: `genConsumeReg` marks the consumed register as not a GC pointer, as it assumes that the input registers
+    // die at the first instruction generated by the node. This is not the case for `INDEX_ADDR`, however, as the
+    // base register is multiply-used. As such, we need to mark the base regsiter as containing a GC pointer until
```
Comment (CarolEidt) [inline src/jit/codegenarmarch.cpp]: Typo: "regsiter"

File: src/jit/codegenxarch.cpp [hunk 6]
```diff
@@ -4495,6 +4499,97 @@ void CodeGen::genCodeForStoreLclVar(GenTreeLclVar* tree)
...
+{
+    GenTree* const base  = node->Arr();
+    GenTree* const index = node->Index();
+
+    genConsumeReg(base);
+    genConsumeReg(index);
+
+    // NOTE: `genConsumeReg` marks the consumed register as not a GC pointer, as it assumes that the input registers
+    // die at the first instruction generated by the node. This is not the case for `INDEX_ADDR`, however, as the
+    // base register is multiply-used. As such, we need to mark the base regsiter as containing a GC pointer until
```
Comment (CarolEidt) [inline src/jit/codegenxarch.cpp]: Same typo as above

**Comment by pgavlin** [comment]
@dotnet-bot test Tizen armel Cross Debug Build

**Pull request merged**